Refactor CompImageHDU to inherit from ImageHDU

astrofrog opened the pull request:
This pull request addresses https://github.com/astropy/astropy/issues/9238 by refactoring ``CompImageHDU`` to inherit from ``ImageHDU`` instead of ``BinTableHDU``. This includes some subtle API changes that cannot be warned about in advance using deprecation warnings or future warnings, so this PR, if accepted, would definitely warrant a major version bump. Ideally this would go in v6.0 but it depends on whether we can achieve consensus and stamp out remaining issues.

This PR isn't 100% ready (see 'Outstanding issues' below and FIXME comments in code) but I wanted to try and open this now in case it is something we can consider getting in before the v6.0 feature freeze. I doubt that the outstanding issues would cause significant code changes so it should be safe to already review things as-is.

### Background

Compressed image HDUs are stored in FITS files as binary tables. The ``CompImageHDU`` was therefore developed as a a subclass of ``BinTableHDU``. However, that implementation is quite complex because it tries to look to users like an ``ImageHDU`` - for example ``.header`` is meant to be the image header, and ``.data`` is meant to be the decompressed image data. To quote comments from the code:

```
# TODO: The difficulty of implementing this screams a need to rewrite this module
```

```
Bypasses `BinTableHDU._writeheader()` which updates the header with
metadata about the data that is meaningless here; another reason
why this class maybe shouldn't inherit directly from BinTableHDU...
```

Issue https://github.com/astropy/astropy/issues/9238 by @mhvk also suggested that ``CompImageHDU`` would be simpler if it inherited from ``ImageHDU``.

This PR does that refactoring, and represents ``CompImageHDU`` as an ``ImageHDU`` subclass that internally stores a ``BinTableHDU`` representing the compressed data. Instead of always keeping the image and table header in sync, headers are only converted at I/O time which should in principle make things faster.

Overall, this results in a few hundred lines of code being removed. For now ``CompImageHeader`` is kept to avoid breaking API so that warnings about reserved keywords are emitted as soon as they are set. However, if we would be happy to 'break' this behavior and only warn at I/O time or when calling ``verify``, we could get rid of the custom header subclass. If we want to keep the behavior, we can nevertheless try and avoid code duplication with the ``Header`` class by ssimplifying the header validation (see https://github.com/astropy/astropy/pull/15293).

### API changes/fixes

I have tried to minimize API changes as much as possible. However, it is impossible to not have any changes at all:

* With this PR, a compressed image HDU ``chdu`` will no longer return ``True`` for ``isinstance(chdu, BinTableHDU)``. There's no way to avoid this and also no way to do this with a gradual deprecation. However, I would hope that this is rare and that if anything people would do ``isinstance(chdu, CompImageHDU)`` which will still be ``True``.

* The following options have changed defaults in the ``CompImageHDU`` initializer to match ``ImageHDU``:

    * ``uint=True`` replaces ``uint=False`` (I need to check if I can come up with an example that shows the difference here)
    * ``scale_back=None`` replaces ``scale_back=False`` (in practice I think this won't change anything)

* The checksum on the main header is now the image checksum not the binary table checksum (I believe this was a bug: https://github.com/astropy/astropy/issues/14753)

### Modified tests

I've tried to change tests as little as possible but there were some unavoidable changes:

* Many tests (probably non-optimally) were checking the contents of ``chdu._header`` which no longer exists (the equivalent is ``chdu._bintable.header``.
* ``test_compressed_header_missing_znaxis`` has been removed - this checked that removing ``ZNAXIS`` or ``ZBITPIX`` from the internal binary table caused accessing the compressed data property to fail. This doesn't seem needed because it relies on modifying private attributes on ``CompImageHDU`` and we now also generate the compressed header on-the-fly so there is no opportunity to remove the keywords before the data is compressed.
* ``test_compressed_header_double_extname`` - included some lines checking state of internal/private properties. Those checks have been removed as the test is no longer relevant for this API.

### Outstanding issues/open questions

* [ ] Add back support for accessing ``CompImageHDU.shape`` (should be simple)
* [ ] Make sure that we don't write out anything when in ``mode='update'`` if a file is simply opened and closed (``test_open_comp_image_in_update_mode`` failure)
* [ ] Not critical for this PR, but currently this preserves the ``CompImageHeader`` class - as noted higher up we could potentially consider removing or simplifying this in future.
* [ ] This deprecates but does not yet remove the ``compressed_data`` property on ``CompImageHDU``. We could keep it but it is probably better to encourage people to open the file with ``disable_image_compression``? I doubt this property is used much.

mhvk committed:
Support set-like operations on Masked values.

mhvk commented on astropy/io/fits/hdu/compressed.py:
```diff
@@ -675,58 +534,109 @@ def __init__(
-        self._orig_bzero = self._bzero
-        self._orig_bscale = self._bscale
-        self._orig_bitpix = self._bitpix
+    @property
```
Is this redefined to avoid inheriting the setter?

astrofrog replied:
I can't remember - but there's a lot of hacks/workarounds currently in this PR, I will go through systematically to make sure everything is needed and will add comments in places.

saimn committed:
improve the MEF section in docs and remove old example

astrofrog committed:
Split out code that does data rescaling on a specified raw data in ImageHDU into a separate method to make it easier to use from subclasses

astrofrog committed:
Added a new _CompBinTableHDU class that will be used to store the compressed image binary table

astrofrog committed:
Refactor CompImageHDU initializer

astrofrog committed:
Remove unecessary CompImageHDU.name property

astrofrog committed:
Add compression_type property

astrofrog committed:
Rename self to bintable in _update_header_data ahead of refactor

astrofrog committed:
Refactor _update_header_data as _get_bintable_without_data

astrofrog committed:
Remove unused _scale_data method

astrofrog committed:
Refactor data property

astrofrog committed:
Simplify compressed_data property

astrofrog committed:
Remove shape property as can use inherited one

astrofrog committed:
Refactor header property as a method that will convert the binary table header to the image header

astrofrog committed:
Update storage of compressed data

astrofrog committed:
Refactor I/O methods

astrofrog committed:
Remove _dtype_for_bitpix and _update_header_scale_info which are no longer needed since we can now rely on the ones on the parent ImageHDU class

astrofrog committed:
Remove old tile_shape and compression_type properties

astrofrog committed:
Add methods and properties required for checksum to work properly

astrofrog committed:
Updated CompImageSection

astrofrog committed:
Updated HDUList to correctly create CompImageHDUs as needed

astrofrog committed:
Adjusted fitsheader script to use internal binary table header

astrofrog committed:
Remove unusued imports

astrofrog committed:
Simplify CompImageHeader but keep it to still check for reserved keywords

astrofrog committed:
Updated tests for compression failures

astrofrog committed:
Updated checksum tests

astrofrog committed:
Fix uint support

astrofrog committed:
Updated CompImageHDU tests

Cadair commented on astropy/io/fits/_tiled_compression/tiled_compression.py:
```diff
@@ -526,6 +526,13 @@ def compress_image_data(
     for irow, tile_slices in _iter_array_tiles(data_shape, tile_shape):
         tile_data = image_data[tile_slices]
 
+        # FIXME: hack for now
```
:rofl:

astrofrog committed:
Fixed more tests

demichieldewilde committed:
BUG: Fix "WinError 1455 the paging file is too small"

This is for large files on windows 32.

nstarman committed:
Fix Parameter tests using all_cls_vars

Signed-off-by: nstarman <nstarman@users.noreply.github.com>

nstarman committed:
Make meta a dataclass field

Signed-off-by: nstarman <nstarman@users.noreply.github.com>

nstarman committed:
Use post-init

Signed-off-by: nstarman <nstarman@users.noreply.github.com>

nstarman committed:
make Parameter compatible with dynamic init

Signed-off-by: nstarman <nstarman@users.noreply.github.com>

nstarman committed:
use dataclass init

Signed-off-by: nstarman <nstarman@users.noreply.github.com>

Signed-off-by: nstarman <nstarman@users.noreply.github.com>

nstarman committed:
refactor private method _init_signature

Signed-off-by: nstarman <nstarman@users.noreply.github.com>

nstarman committed:
Freeze Cosmology dataclass

Signed-off-by: nstarman <nstarman@users.noreply.github.com>

nstarman committed:
Freeze Cosmology dataclass

Signed-off-by: nstarman <nstarman@users.noreply.github.com>

nstarman committed:
use dataclass repr

Signed-off-by: nstarman <nstarman@users.noreply.github.com>

nstarman committed:
Simplify Parameter property creation

Signed-off-by: nstarman <nstarman@users.noreply.github.com>

mhvk committed:
DOC: describe how to create test venv with tox

pllim committed:
Backport PR #15954: RTD: Fix astropy version in v6.0.x doc build

neutrinoceros committed:
TST: add a regression test for PR 15956

neutrinoceros committed:
BUG: fix an incorrect error message in io.votable.TableElement.parse

Co-authored-by: Eero Vaher <eero.vaher@gmail.com>

neutrinoceros committed:
PERF: optimize a hot path in Angle.__new__

neutrinoceros committed:
TST: Mark slow tests in astropy/samp

mhvk committed:
BUG: ensure gufuncs using structured dtypes can work with Masked

mhvk committed:
BUG: ensure find_equivalent_units returns proper equivalent units

MridulS committed:
MAINT: Let ruff format run on docs and examples (#16132)

* MAINT: Let ruff format run on docs and examples

mhvk committed:
TST: check that masking invalid inputs prevents erfa warnings/exceptions

mhvk committed:
Merge pull request #16125 from mhvk/masked-fix-structured-dtype-ufunc

BUG: Ensure gufuncs using structured dtypes can work with Masked

sbailey committed:
test and doc updates

mhvk committed:
Merge pull request #16130 from sbailey/table_vstack_inputs

table.vstack shouldn't modify inputs

nstarman committed:
feat: fast-path physicsspherical to cylindrical

Signed-off-by: nstarman <nstarman@users.noreply.github.com>

nstarman committed:
docs: changelog

Signed-off-by: nstarman <nstarman@users.noreply.github.com>

astrofrog committed:
Implemented improvements suggested by @mhvk

eerovaher committed:
Shorten `_get_frame_without_data()`

mhvk committed:
Merge pull request #16137 from eerovaher/shorten_get_frame_without_data

Shorten `_get_frame_without_data()`

Commit:
Update minimum required version of astropy-iers-data

nstarman committed:
add cylindrical -> physicsspherical

nstarman committed:
add tests

neutrinoceros committed:
BUG: declare private str ufuncs as unsupported (np._core.umath._expandtabs*)

mhvk committed:
Merge pull request #16144 from neutrinoceros/numpy2/new_str_ufuncs

BUG: declare private str ufuncs as unsupported (np._core.umath._expandtabs*)

nstarman committed:
Delete docs/changes/coordinates/16135.feature.rst

astrofrog committed:
Fix code style

nstarman committed:
test: allclose

Signed-off-by: nstarman <nstarman@users.noreply.github.com>

neutrinoceros committed:
TST: temporarily pin pytest<8.1

MridulS committed:
MAINT: Use array-api-strict instead of np.array_api

mhvk committed:
Merge pull request #16147 from neutrinoceros/tests/pin_pytest

TST: temporarily pin pytest<8.1

astrofrog committed:
Remove unused code

astrofrog committed:
Don't allow _data_offset, _header_offset, and _data_size to be set to non-None values.

astrofrog committed:
Updated fitsio-dependent test to not access private header

pllim committed:
Revert "TST: temporarily pin pytest<8.1"

pllim committed:
Merge pull request #16140 from astropy/update-astropy-iers-data-pin-1709338155

Update minimum required version of astropy-iers-data

mhvk committed:
Merge pull request #16155 from astropy/revert-16147-tests/pin_pytest

Revert "TST: temporarily pin pytest<8.1" - pytest 8.1 has been yanked
because it caused so much trouble...

neutrinoceros committed:
BUG: fix compatibility with numpy 2.0 copy semantics

neutrinoceros committed:
TST: fix a broken test

mhvk committed:
Let Quantity use COPY_IF_NEEDED internally

mhvk committed:
Merge pull request #16150 from MridulS/use_array-api-strict

MAINT: Use array-api-strict instead of np.array_api

mhvk committed:
Merge pull request #16142 from neutrinoceros/numpy2/array_copy_False

BUG: fix compatibility with numpy 2.0 copy semantics

lpsinger committed:
BUG: Fix loading parser tabs on pyc-only installations

According to the Python module loading
[flow chart](https://peps.python.org/pep-3147/#flow-chart), when
evaluating `import foo` and `foo.py` is not found, Python then
reads `foo.pyc`.

One can take advantage of this fact to strip source files and
leave only Python bytecode files for deployment in
space-constrained execution environments such as AWS Lambda
(see, for example, nasa-gcn/gcn.nasa.gov#2040).

Fix a bug in the wrappers for the lex and yacc wrappers that are
used for parsing Astropy units so that they work on pyc-only
installations.

neutrinoceros committed:
API: align default value for copy arguments with numpy at runtime

manodeep committed:
Replaced pow with a multiply

neutrinoceros committed:
RFC: refactor copy argument sanitizing into a unified compat function

larrybradley committed:
Merge pull request #16169 from manodeep/fix_pow

Replaced pow with a multiply

mhvk committed:
Merge pull request #16159 from lpsinger/lex-pyc

BUG: Fix loading parser tabs on pyc-only installations

eerovaher committed:
Replace a private function with inline lambdas

Code that `astropy.coordinates.representation.base._make_getter()` is
replaced with is so simple that there is no need to keep the function.

mhvk committed:
Merge pull request #16173 from eerovaher/rm-make-getter

Remove `astropy.coordinates.representation.base._make_getter()`

pllim committed:
DOC: Add test runner limitation wrt asdf-astropy to Known Issues (#16179)

* DOC: Add test runner limitation
wrt asdf-astropy to Known Issues. [ci skip]

* [pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

---------

Co-authored-by: pre-commit-ci[bot] <66853113+pre-commit-ci[bot]@users.noreply.github.com>

neutrinoceros committed:
API: change meaning of copy=False in astropy APIs (align with runtime numpy)

namurphy committed:
DOC: Speed up Sphinx build by using `sphinxcontrib-globalsubs` (#16162)

* Apply sphinxcontrib-globalsubs to substitutions across the package

* Remove unused substitutions/links

* Add comments to conf.py

* Move over common_links.rst to conf.py

pllim committed:
Merge pull request #15603 from nstarman/bump-py-310

Bump minimum Python to 3.10

neutrinoceros committed:
TST: remove workarounds now that contourpy has unpinned nightlies (#16122)

* TST: remove workarounds now that contourpy has unpinned nightlies

* TST: Recombine mpldev into devdeps CI job while retaining mpldev for image tests.

---------

Co-authored-by: P. L. Lim <2090236+pllim@users.noreply.github.com>

mhvk committed:
Merge pull request #16170 from neutrinoceros/rfc/sanitize_copy_arg

RFC: refactor copy argument sanitizing into a unified compat function

nstarman committed:
Simplify cloning

Signed-off-by: nstarman <nstarman@users.noreply.github.com>

nstarman committed:
Make name a dataclass field

py3.13+ will have the `converter` arg, eliminating this class.

Co-authored-by: Clément Robert <cr52@protonmail.com>
Signed-off-by: nstarman <nstarman@users.noreply.github.com>

neutrinoceros committed:
RFC: refactor internal usage of copy=False to copy=COPY_IF_NEEDED (#16166)

Now internally we will be OK regardless of how we deal with the numpy 2.0
change in the meaning of `copy`.

neutrinoceros committed:
CLN: cleanup unused bits now that Python 3.9 was dropped

neutrinoceros committed:
CLN: cleanup unused try block in astropy/convolution/convolve.py

eerovaher committed:
Remove caching from a function that is not called

`TransformGraph.frame_component_names()` is not used by `astropy`, nor
is it expected to be frequently called by users, so caching complicates
it without providing any benefits.

eerovaher committed:
Refactor two `TransformGraph` methods

Both methods implemented caching by carefully managing an associated
private variable, but now they use `astropy.utils.lazyproperty` instead.
This is preferred over using `functools.cached_property` because if the
cached property was not accessed then clearing the cache makes the
latter raise an `AttributeError` that needs to be suppressed explicitly.

larrybradley committed:
Merge pull request #16184 from neutrinoceros/mnt/cleanup_unnecessary_try_block

CLN: cleanup unused try block in astropy/convolution/convolve.py

mhvk committed:
Merge pull request #16182 from eerovaher/transformation-graph-caching

Simplify coordinate transformation graph caching

pllim committed:
DOC: Test guide should be consistent
with actual jobs we run. [ci skip]

pllim committed:
Merge pull request #16185 from pllim/test-doc-py39

DOC: Test guide should be consistent with actual jobs we run

mhvk committed:
Merge pull request #16183 from neutrinoceros/mnt/cleanup_post_3.10_adoption

CLN: cleanup unused bits now that Python 3.9 was dropped

neutrinoceros committed:
RFC: inline usage of sys.version_info to enable automated clean up of compatibility code

nstarman committed:
address review comments

Signed-off-by: nstarman <nstarman@users.noreply.github.com>

astrofrog committed:
Fix remaining test related to writing out files in update mode when data has not been modified

astrofrog committed:
Remove _summary as can use inherited method

astrofrog committed:
Fix code style

astrofrog committed:
Added a test of HDUList.info() when compressed image HDU is present

astrofrog committed:
Updated documentation

astrofrog committed:
Added a test for CompImageHDU.shape

astrofrog committed:
Fix test where data shape was modified on-the-fly by the user

astrofrog committed:
Avoid unecessary data updates, only do it in the case where the shape has changed but the data object has not

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

astrofrog committed:
Remove comment

astrofrog committed:
Added a test of the uint= option

astrofrog committed:
Added regression tests for known CompImageHDU bugs

pllim committed:
Merge pull request #16186 from neutrinoceros/rfc/auto_upgradable_compat_code

RFC: inline usage of sys.version_info to enable automated clean up of compatibility code

astrofrog committed:
Fix regression tests and implementation

astrofrog committed:
Added changelog entries

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

astrofrog committed:
Fix decompressed image header

astrofrog committed:
Preserve BSCALE/BZERO during header conversion

eerovaher committed:
Implement `Table.setdefault()`

The new method is analogous to `dict.setdefault()`.

pllim committed:
Merge pull request #16189 from meeseeksmachine/auto-backport-of-pr-15954-on-main

Backport PR #15954 on branch main (RTD: Fix astropy version in v6.0.x doc build)

pllim committed:
Merge pull request #16095 from neutrinoceros/tests/rfc/mark_slow_tests

TST: Mark slow tests in astropy/samp

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

jdavies-st committed:
Sort test parametrization in cosmology

taldcroft committed:
Merge pull request #16188 from eerovaher/table-setdefault

pllim committed:
Merge pull request #16191 from jdavies-st/testing-cosmology-fix-unsorted-parametrization

TST: Sort test parametrization in cosmology

pllim committed:
DOC: Move FITS example gallery to FAQ

eerovaher committed:
Update Ruff `pre-commit` hook

The hook is updated manually to apply three additional manual formatting
changes: one in astropy/table/tests/test_column.py and two in
astropy/visualization/tests/test_time.py

nstarman committed:
Make Cosmology a dataclass

Signed-off-by: nstarman <nstarman@users.noreply.github.com>

Signed-off-by: nstarman <nstarman@users.noreply.github.com>

jdavies-st committed:
TST: Speed-up test suite when using pytest-xdist (#16196)

* Use threadpoolctl to limit threads when pytest-xdist in use

* Use threadpoolctl in pytest_configure hook

* Use try/except/else

pllim committed:
DOC: Remove os.remove in FITS FAQ.

Co-authored-by: Simon Conseil <contact@saimon.org>

neutrinoceros committed:
BUG/API: Changed the default value of the copy argument in CylindricalDifferential.__init__ from False to True

mhvk committed:
Merge pull request #16198 from neutrinoceros/coordinates/bug/api/CylindricalDifferential_copy_by_default

BUG/API: Changed the default value of the copy argument in CylindricalDifferential.__init__ from False to True

eerovaher committed:
Remove `astropy.utils.diff.fixed_width_indent()`

It is better to use `textwrap.indent()` from the Python standard library
instead.

ayshih committed:
Improve automatically retrieved documentation of coordinate-frame attributes (#15259)

Return the descriptor instance rather than its default value when accessed from a class
and allow a customizable docstring for each frame attribute, appending information
on the default value to the docstring.

eerovaher committed:
Deprecate dangerous `TableColumns` methods

`TableColumns` inherits its `setdefault()` and `update()` methods from
`OrderedDict`, which is not aware of `Table`. Using those methods
therefore bypasses normal table column validation and conversion
mechanisms and can easily lead to `Table` corruption. It is not possible
to fix them entirely because the `TableColumns` instance attached to a
`Table` without any columns has no way of knowing what `Table` instance
it is attached to, so it has no way of ensuring proper column creation
takes place. Although this is a single edge case, it is not negligible.
It is therefore best if the methods raise an error before they have a
chance of corrupting the table, but they need to be deprecated first.

mbtaylor committed:
Update SAMP hub web profile for Private Network Access (#16193)

* Update SAMP hub web profile for Private Network Access

The CORS HTTP server handling localhost web profile requests
now does header manipulation as required by the Private Network
Access proposal (https://wicg.github.io/private-network-access/).
At present, the effect is to prevent recent versions of Google Chrome
(e.g. 122) from complaining "Ensure private network requests are
only made to resources that allow them"; future versions of
Chrome or possibly other browsers may block such requests.

* add change note

* Reformat link in change log

---------

Co-authored-by: Tom <tdonaldson@stsci.edu>

mhvk committed:
Add pytest links and mention --last-failed option [skip ci]

mhvk committed:
Merge pull request #15973 from eerovaher/fixed_width_indent

Remove `fixed_width_indent()`

saimn committed:
improve unified io section and remove chandra example [docs only]

MridulS committed:
DOC: pandas Series __mul__ was updated in 2.0

pllim committed:
Merge pull request #16202 from MridulS/MridulS-doc-2

DOC: pandas Series __mul__ was updated in 2.0

pllim committed:
DOC: Un-ignore OSU link 

now that they finally fixed the SSL cert

MridulS committed:
ENH: Run asv benchmarks on demand for PRs

pllim committed:
Merge pull request #16204 from pllim/unignore-osu-ssl

DOC: Un-ignore OSU link in linkcheck

pllim committed:
Merge pull request #15779 from MridulS/move-benchmarks

Add continuous CI benchmarking using astropy-benchmarks

neutrinoceros committed:
BUG: declare private str ufuncs as unsupported (np._core.umath._(ljust, rjust, _center, _zfill))

mhvk committed:
Merge pull request #16205 from neutrinoceros/numpy2/new_str_ufuncs_2

BUG: declare private str ufuncs as unsupported (np._core.umath._(ljust, rjust, _center, _zfill))

mhvk committed:
Merge pull request #16127 from mhvk/units-find-equivalent-units-fix

BUG: ensure find_equivalent_units returns proper equivalent units

saimn committed:
move plot example to usage/image.rst

saimn committed:
fix chandra example

saimn committed:
improve convenience section

pllim committed:
Merge pull request #15878 from mhvk/devdocs-testguide-editable-install

DOC: describe how to create a test environment with tox

saimn committed:
Fix SyntaxWarning: invalid escape sequence in docs build

neutrinoceros committed:
DEPR: deprecate Time.ptp in favor of np.ptp with Numpy >= 2.0

pllim committed:
Merge pull request #16213 from saimn/docs-syntaxerror

Fix SyntaxWarning: invalid escape sequence in docs build

saimn committed:
remove plot example

eerovaher committed:
Merge pull request #16088 from neutrinoceros/coordinates/perf/optimize_hotpath_Angle__new__

PERF: optimize a hot path in `Angle.__new__`

pllim committed:
Merge pull request #16194 from pllim/exorcise-fits-example-gallery

DOC: Move FITS example gallery to FAQ

neutrinoceros committed:
PERF: reduce validation overhead in `Latitude._validate_angles` (#16096)

mhvk committed:
Allow unmasked scalar Masked instances to be hashed

mhvk committed:
Regression test for SkyCoord.apply_space_motion

ViciousEagle03 committed:
Fix issue in ndcube requiring changes in astropy (#15985)

* Fix issue in ndcube requiring changes in astropy

* Fix issue in ndcube (required changes in astropy)Handled the case of  separately and  separately

* Fix issue in ndcube (required changes in astropy)

* Fix issue in ndcube (required changes in astropy)

* Restructured the patch

* Add changlog and regression test

* Update docs/changes/visualization/15985.bugfix.rst

Co-authored-by: Clément Robert <cr52@protonmail.com>

* Corrected edge case

* Modified tests for my patch

* [pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

* Update docs/changes/visualization/15985.bugfix.rst

Co-authored-by: Clément Robert <cr52@protonmail.com>

* Removed the case where draw_grid value is invalid

* Update docs/changes/visualization/15985.bugfix.rst

Co-authored-by: P. L. Lim <2090236+pllim@users.noreply.github.com>

---------

Co-authored-by: Clément Robert <cr52@protonmail.com>
Co-authored-by: pre-commit-ci[bot] <66853113+pre-commit-ci[bot]@users.noreply.github.com>
Co-authored-by: P. L. Lim <2090236+pllim@users.noreply.github.com>
Co-authored-by: Stuart Mumford <stuart@cadair.com>

eerovaher committed:
Reorganize coordinate separation tests

The tests for `BaseCoordinateFrame` `position_angle()`, `separation()`
and `separation_3d()` methods are inconveniently scattered accross many
files. Furthermore, `SkyCoord` instances expose the separation methods
of their underlying frame, and tests for that functionality are also
scattered accross many files. It is good to collect the tests in a
single file where they can be parametrized over the coordinate class
being either `SkyCoord` or a `BaseCoordinateFrame` subclass, and the
method input can be parametrized likewise. By introducing a simple
helper for storing the expected test outcomes it becomes easy to
parametrize the tests over the three separation methods too.

There have been cases where a bug has been identified and fixed in a
`SkyCoord` method, but not in the analogous `BaseCoordinateFrame`
method. The new parametrized tests should prevent that.

mhvk committed:
Merge pull request #16215 from eerovaher/coord-separation-tests

Reorganize coordinate separation method tests

neutrinoceros committed:
PERF: skip angle wrapping when possible

mhvk committed:
Merge pull request #16222 from neutrinoceros/coordinates/perf/skip_wrapping_when_possible

PERF: skip angle wrapping when possible

mhvk committed:
Merge pull request #16212 from neutrinoceros/time/depr/ptp

DEPR: deprecate Time.ptp in favor of np.ptp with Numpy >= 2.0

eerovaher committed:
Deprecate `astropy.utils.misc.indent()`

The function was made obsolete by `textwrap.indent()`, introduced in
Python version 3.3.

pllim committed:
Merge pull request #16223 from eerovaher/deprecate-indent

Deprecate `astropy.utils.misc.indent()`

mhvk committed:
Adjust time tests for unmasked Masked scalar being hashable

eerovaher committed:
Shorten `_parse_coordinate_data()`

nstarman committed:
add tests for the origin

Signed-off-by: nstarman <nstarman@users.noreply.github.com>

MridulS committed:
MAINT: Benchmark action should fail if a regression is caught (#16207)

* MAINT: Benchmark action should fail if a regression is caught

* Make sure upload always run

pllim committed:
Merge pull request #16226 from eerovaher/shorter_parse_coordinate_data

Shorten `_parse_coordinate_data()`

eerovaher committed:
Remove `_parse_coordinate_arg()` docstring

It is clearly not describing the actual function.

eerovaher committed:
Simplify `_parse_coordinate_arg()` signature

The `init_kwargs` argument was used in one line where it was unpacked,
but `_parse_coordinate_arg()` is only called in one place where it is
guaranteed that `init_kwargs` could only be an empty `dict`.

mhvk committed:
Merge pull request #16160 from eerovaher/ruff-030

Manual `pre-commit` update

mhvk committed:
Merge pull request #16228 from eerovaher/_parse_coordinate_arg_signature

Simplify `_parse_coordinate_arg()` signature

MridulS committed:
ENH: Move _get_converter to public API

neutrinoceros committed:
BUG: fix compatibility with numpy@maintenance/v2.0.x

taldcroft committed:
Merge pull request #16154 from eerovaher/tablecolumns-setdefault

Deprecate `TableColumns.setdefault()` and `TableColumns.update()`

mhvk committed:
Merge pull request #16229 from neutrinoceros/numpy2/adjust_units_helper

BUG: fix compatibility with numpy@maintenance/v2.0.x

mhvk committed:
Merge pull request #16139 from MridulS/get_converter

ENH: Move _get_converter to public API

neutrinoceros committed:
BUG: fix calling Table.group_by on an empty table (#16093)

* TST: add a regression test for bug 11884

* BUG: fix calling Table.group_by on an empty table

olintoeduardo committed:
Removed OrderedDict from time in formats.py (#15491)

* Update formats.py

Removed OrderedDict from astropy/time/formats.py and replaced with dict.

* Create 15491.api.rst

* Change comment

[ci skip]

* Reword change log

* Use dict literal

---------

Co-authored-by: P. L. Lim <2090236+pllim@users.noreply.github.com>
Co-authored-by: Tom Aldcroft <taldcroft@gmail.com>

eerovaher committed:
Add type annotations to introspection functions

A Sphinx bug requires the documentation configuration to be updated too.

neutrinoceros committed:
TST: add a regression test for issue 16219

neutrinoceros committed:
BUG: fix a bug where SkyCoord.spherical_offsets_by would crash when a wrap was needed

neutrinoceros committed:
DOC: add a note that wrapping Angle with integral dtypes isn't supported (#16232)

mhvk committed:
Also fix problems for non-matching size

eerovaher committed:
Speed up `Angle` creation

`Angle._convert_unit_to_angle_unit()` is cached, so it is better to move
as much code as possible inside it. The speed-up is especially large if
the `Angle` unit is specified as a `str`.

mhvk committed:
Merge pull request #16243 from eerovaher/faster-angle-unit-converter

Speed up `Angle` creation

neutrinoceros committed:
DOC: rework documentation for the __astropy_table__ protocol to avoid mingling with numpy 2.0's copy semantics (#16239)

nstarman committed:
Apply suggestions from code review

Co-authored-by: Eero Vaher <eero.vaher@gmail.com>

mhvk committed:
Add changelog entry and test case

And slightly modify fix, to check that "winerror" is in fact present.

neutrinoceros committed:
BUG: fix a bug where columns with dtype=object wouldn't be properly deep-copied using copy.deepcopy (#15871)

* TST: add regression tests for bug 13435

* BUG: fix a bug where columns with dtype=object wouldn't be properly deep-copied using copy.deepcopy

mhvk committed:
Merge pull request #16135 from nstarman/coords-pysph-to-cyl

feat: fast-path physicsspherical to cylindrical

nstarman committed:
Merge pull request #16224 from mhvk/masked-setops-functions

Support Masked setops functions and hashing of scalars.

eerovaher committed:
Make computing angular separations less surprising

When computing the angular separation between coordinates in different
frames they must first be transformed into a common frame. However, the
resulting angular separation value can depend on the direction of the
transformation, unless it is a pure rotation. Users have found this to
be surprising, so by default a warning is now emitted if a non-rotation
transformation is needed. It is possible to always suppress the warning,
or to forbid non-rotation transformations.

Computing the position angle always performs the transformation silently
because its value is expected to depend on the order of the coordinates
anyways.

nstarman committed:
Make FLRW et al dataclasses

Signed-off-by: nstarman <nstarman@users.noreply.github.com>

Signed-off-by: nstarman <nstarman@users.noreply.github.com>

eerovaher committed:
Merge pull request #15484 from nstarman/cosmo-cosmo-dataclass

Cosmo as a dataclass

eerovaher committed:
Add What's New entry for separation `origin_mismatch`

mhvk committed:
Merge pull request #16246 from eerovaher/separation-frame-mismatch

Make computing angular separations less surprising

MridulS committed:
DOCS: Add docs about CI benchmarking (#16225)

* DOCS: Add docs about CI benchmarking [ci skip]

* add docs

* Apply suggestions from code review

Co-authored-by: P. L. Lim <2090236+pllim@users.noreply.github.com>

* Apply suggestions from code review

Co-authored-by: Eero Vaher <eero.vaher@gmail.com>
Co-authored-by: Clément Robert <cr52@protonmail.com>

---------

Co-authored-by: P. L. Lim <2090236+pllim@users.noreply.github.com>
Co-authored-by: Eero Vaher <eero.vaher@gmail.com>
Co-authored-by: Clément Robert <cr52@protonmail.com>

MridulS committed:
FIX: preserve round trip with parquet i/o (#16237)

* FIX: preserve round trip with parquet i/o

* add changelog

invalid-email-address committed:
Update minimum required version of astropy-iers-data

mhvk committed:
Merge pull request #16256 from astropy/update-astropy-iers-data-pin-1712016580

Update minimum required version of astropy-iers-data

eerovaher committed:
Prefer `scipy.spatial` `KDTree` over `cKDTree`

Before `scipy` v1.6.0 it used to be the case that `cKDTree` was more
performant, but since then they are functionally identical and the
recommendation is to prefer `KDTree`.

mhvk committed:
Merge pull request #16259 from eerovaher/rm-cKDTree

Prefer `scipy.spatial` `KDTree` over `cKDTree`

pllim committed:
Merge pull request #15956 from neutrinoceros/io/votable/bug/only_report_missing_columns

BUG: fix an incorrect error message in io.votable.TableElement.parse

neutrinoceros committed:
TST: add a regression test for issue 16260

neutrinoceros committed:
BUG: fix a regression in astropy 6.0.1 where using a dimensionless Quantity as an exponent could raise an exception

mhvk committed:
Merge pull request #16241 from neutrinoceros/coordinates/bug/SkyCoord_spherical_offsets_by_with_wrap_16219

BUG: fix a bug where `SkyCoord.spherical_offsets_by` would crash when a wrap was needed

neutrinoceros committed:
BLD: build with numpy 2.0.0rc1 (or newer) (#16252)

* BLD: build with numpy 2.0.0rc1 (or newer)

* DOC: add changelog entry

pllim committed:
Merge pull request #16234 from eerovaher/annotate-introspection

Add type annotations to introspection functions

henryiii committed:
chore: versions are numbers

See #16253.

pllim committed:
Merge pull request #16263 from henryiii/patch-1

chore: versions are numbers

svank committed:
A couple of WCS DistortionLookupTable fixes (#16163)

* Add tests for WCSes with DistortionLookupTable

* Fix distortion lookups from DistortionLookupTable

In image_coord_to_distortion_coord, the input coordinates (in the image
plane) and crval/crpix are all 1-based pixel coordinates, so the
conversion to 0-based pixel coordinates, which is happening in the
distortion-table plane, doesn't need a scale factor.

* Fix slicing of WCS with DistortionLookupTables

* Add changelog

eerovaher committed:
Update Ruff `pre-commit` hook to v0.3.4

Ruff rule UP032 can automatically convert `str.format()` calls to
f-strings. Before v0.3.2 Ruff avoided making changes if the naive
conversion produced lines that exceeded the line length limit, but now
it is more aggressive. The `pre-commit` hook is updated manually to
prevent Ruff from producing overlong lines.

saimn committed:
Merge pull request #15388 from demichieldewilde/winerror1455fix

'BUG: WinError 1455 the paging file is too small'

neutrinoceros committed:
Update 16181.other.rst

Co-authored-by: P. L. Lim <2090236+pllim@users.noreply.github.com>

mhvk committed:
Merge pull request #16261 from neutrinoceros/units/bug/quantity_as_power_16260

BUG: fix a regression in astropy 6.0.1 where using a dimensionless Quantity as an exponent could raise an exception

neutrinoceros committed:
RFC: use actual numpy dev version numbers in NUMPY_LT_* definitions

neutrinoceros committed:
BUG: declare private str ufuncs as unsupported (np.core.math._(r)partition(_index))

astrofrog committed:
Merge pull request #16181 from neutrinoceros/api/copy_if_needed_clean_break

API: align default value for copy arguments with numpy at runtime (copy=False raises an error if copies cannot be avoided)

astrofrog committed:
Merge pull request #16270 from neutrinoceros/units/bug/numpy_2_1_new_string_ufuncs

BUG: declare private str ufuncs as unsupported (np.core.math._(r)partition(_index))

astrofrog committed:
Merge pull request #16264 from eerovaher/ruff-034

Update Ruff `pre-commit` hook to v0.3.4

astrofrog committed:
Added 7.0 what's new page and redirect 6.1 what's new page

astrofrog committed:
Updated list of contributors and .mailmap file

maxnoe committed:
Add --color to sphinx linkcheck in ci

astrofrog committed:
Updated what's new link

Cadair committed:
Remove Cadair from codeowners

Too many linting PRs

pllim committed:
Merge pull request #16275 from Cadair/no_codeowner_cadair

Remove Cadair from codeowners

mhvk committed:
Merge pull request #16269 from neutrinoceros/mnt/numpy_version_comp

RFC: use actual numpy dev version numbers in NUMPY_LT_* definitions

pllim committed:
Merge pull request #16273 from maxnoe/linkcheck-color

Add --color to sphinx linkcheck in ci

eerovaher committed:
Prefer module setup over function setup in a test file

All the coordinate matching tests require `scipy` to be installed, so
the entire module can be skipped if it is not. Furthermore, it is better
to import at the module level than in each test function separately.

eerovaher committed:
Rewrite an overlong coordinate matching test

Smaller parametrized tests allow more checks to run if one of them
fails.

pllim committed:
Merge pull request #16277 from eerovaher/coord-matching-tests

Rewrite an overlong coordinate matching test function

saimn committed:
Update .mailmap

saimn committed:
Merge pull request #16272 from astrofrog/v6.1.0-mailmap-credits

Updated list of contributors and .mailmap file

astrofrog committed:
Always run verify() on decompressed image header

astrofrog committed:
Try and fix _verify for cases where the decompressed header is a primary header not an extension header.

astrofrog commented:
Closing as the astropy PR is now open.
